torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -449,6 +474,90 @@ void ProcessGroupAgent::listenLoop() {
   }
 }
 
+void ProcessGroupAgent::pollTimedOutRPCs() {
+  while (!shutdown_.load()) {
+    std::chrono::milliseconds sleepTime;
+    {
+      std::unique_lock<std::mutex> lock{futureMutex_};
+      // Estimate amount of time the first future will time out in, and sleep
+      // for that long.
+      // if there are no futures or the RPC timeout is set to 0 (meaning no
+      // timeout), then sleep for a set "infinity" time.
+      if (futureTimeouts_.empty() || rpcTimeout_.count() == 0) {
+        sleepTime = INFINITY_SLEEP_INTERVAL;
+      } else {
+        auto runningTime =
+            std::chrono::duration_cast<std::chrono::milliseconds>(
+                std::chrono::steady_clock::now().time_since_epoch()) -
+            futureTimeouts_.begin()->first;
+        sleepTime =
+            std::max(rpcTimeout_ - runningTime, std::chrono::milliseconds(0));
+      }
+      futureTimeoutCV_.wait_for(lock, sleepTime);
+    }
+
```

Comment:
Can we add a check here that if `rpcTimeout_.count() == 0` we just continue and go back to sleeping? We don't need to call `processTimedOutFutures` right?

Reply:
But if rpcTimeout_.count() == 0, we're basically rarely/never waking up, or hitting this codepath, right?
(maybe I'm missing something?)

Reply:
This should be the case, since we would run once, sleep for a long time, and wake up when shutting down.

Reply:
Wouldn't the `notify_one` in the send function wake us up every time there is on outstanding RPC?

Reply:
No, notify_one in send() only needs to wake this up when there's a sooner timeout that the present one.
(note that the if size==1 codepath, i.e. it was previously empty)

Reply:
We only call notify on `futureTimeoutCV` when there is the first future or when shutting down.